tests/cases/compiler/tsxInferenceShouldNotYieldAnyOnUnions.tsx

```diff
@@ -0,0 +1,29 @@
+// @jsx: preserve
+// @filename: index.tsx
+namespace JSX {
+    export interface Element {}
+}
+
+type Props<T> = PropsBase<string> | PropsWithConvert<T>;
+
+interface PropsBase<T> {
+    data: T;
+}
+
+interface PropsWithConvert<T> extends PropsBase<T> {
+    convert: (t: T) => string;
+}
+
+function InferenceShouldNotProduceAny<T>(props: Props<T>): JSX.Element {
+    return <div />;
+}
+
+// Sanity check: function call equivalent versions work fine
+InferenceShouldNotProduceAny({ data: "1" });
+InferenceShouldNotProduceAny({ data: "1", convert: n => "" + n });
+InferenceShouldNotProduceAny({ data: 2, convert: n => "" + n });
+
+
+const f1 = <InferenceShouldNotProduceAny data={"1"} />;

```

Comment:
This is tiny, but a positive name like InferenceShouldProduceNumber would be better.